# Исследование надежности заемщиков


Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом. 



## Откройте таблицу и изучите общую информацию о данных

**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`

In [1]:
import pandas as pd

try:
    data = pd.read_csv('/datasets/data.csv')
except:
    data = pd.read_csv('https://code.s3.yandex.net/datasets/data.csv')

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**

In [2]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**

In [3]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных

### Удаление пропусков

**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**

In [4]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**

In [5]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['total_income'].isna()), 'total_income'] = \
    data.loc[(data['income_type'] == t), 'total_income'].median()

### Обработка аномальных значений

**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**

In [6]:
data['days_employed'] = data['days_employed'].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**

In [7]:
data.groupby('income_type')['days_employed'].agg('median')

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть.

**Задание 8. Выведите перечень уникальных значений столбца `children`.**

In [8]:
data['children'].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5])

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**

In [9]:
data = data[(data['children'] != -1) & (data['children'] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**

In [10]:
data['children'].unique()

array([1, 0, 3, 2, 4, 5])

### Удаление пропусков (продолжение)

**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**

In [11]:
for t in data['income_type'].unique():
    data.loc[(data['income_type'] == t) & (data['days_employed'].isna()), 'days_employed'] = \
    data.loc[(data['income_type'] == t), 'days_employed'].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**

In [12]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных

**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**

In [13]:
data['total_income'] = data['total_income'].astype(int)

### Обработка дубликатов

**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру.**

In [14]:
data['education'] = data['education'].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**

In [15]:
data.duplicated().sum()

71

In [16]:
data = data.drop_duplicates()

### Категоризация данных

**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.


**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**

In [17]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return 'E'
        elif 30001 <= income <= 50000:
            return 'D'
        elif 50001 <= income <= 200000:
            return 'C'
        elif 200001 <= income <= 1000000:
            return 'B'
        elif income >= 1000001:
            return 'A'
    except:
        pass

In [18]:
data['total_income_category'] = data['total_income'].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**

In [19]:
data['purpose'].unique()

array(['покупка жилья', 'приобретение автомобиля',
       'дополнительное образование', 'сыграть свадьбу',
       'операции с жильем', 'образование', 'на проведение свадьбы',
       'покупка жилья для семьи', 'покупка недвижимости',
       'покупка коммерческой недвижимости', 'покупка жилой недвижимости',
       'строительство собственной недвижимости', 'недвижимость',
       'строительство недвижимости', 'на покупку подержанного автомобиля',
       'на покупку своего автомобиля',
       'операции с коммерческой недвижимостью',
       'строительство жилой недвижимости', 'жилье',
       'операции со своей недвижимостью', 'автомобили',
       'заняться образованием', 'сделка с подержанным автомобилем',
       'получение образования', 'автомобиль', 'свадьба',
       'получение дополнительного образования', 'покупка своего жилья',
       'операции с недвижимостью', 'получение высшего образования',
       'свой автомобиль', 'сделка с автомобилем',
       'профильное образование', 'высшее об

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**

In [20]:
def categorize_purpose(row):
    try:
        if 'автом' in row:
            return 'операции с автомобилем'
        elif 'жил' in row or 'недвиж' in row:
            return 'операции с недвижимостью'
        elif 'свад' in row:
            return 'проведение свадьбы'
        elif 'образов' in row:
            return 'получение образования'
    except:
        return 'нет категории'

In [21]:
data['purpose_category'] = data['purpose'].apply(categorize_purpose)

### Шаг 3. Исследуйте данные и ответьте на вопросы

#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?

In [22]:
def category_children(row):
    if row == 0:
        return 'нет детей'
    elif row == 1:
        return 'один ребенок'
    elif row == 2:
        return 'два ребенка'
    else:
        return 'многодетная семья'
    
data['quantity_children'] = data['children'].apply(category_children)

svod_children = data.pivot_table(index = 'quantity_children', values = 'debt')
svod_children['quantity_debt'] = data.groupby('quantity_children')['debt'].count()

def percent(x):
    return "{0:.2%}".format(x)

svod_children['debt'] = svod_children['debt'].apply(percent)
svod_children = svod_children.sort_values(by='debt', ascending=False)
svod_children.rename(columns={'debt':'Процент задолженостей', 'quantity_debt':'Количество кредитов группе'})

Процент задолженостей  Количество кредитов группе
quantity_children                                                  
два ребенка                       9.45%                        2052
один ребенок                      9.23%                        4808
многодетная семья                 8.16%                         380
нет детей                         7.54%                       14091

<div class="alert alert-success">

**Артём:** 👍 Молодец, что используешь сводную таблицу. 

</div>

**Вывод:** 

Изначально количество детей в семье было сгруппировано в группу "Многодетная семья" для того, чтобы выделить семьи с 3 и более детьми. Это сделано, чтобы показать наглядно, что количество таких случаев в данных очень мало в сравнении с общим количеством, поэтому эту группу мы не будем брать в учет при расмотрении зависимости.

Из таблицы можно сделать вывод, что семьи без детей меньше всего имеют процент задолженностей 7.54%, и количество таких случаев самое большое 14091, что подтверждает закономерность. Так же чем больше детей, тем больший процент задолженности "один ребенок" - 9.23%, "два ребенка" - 9.45%, исключая случаи многодетных семей 8.16%.


<div class="alert alert-success">

**Артём:** 👍 Верно.

</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

In [23]:
svod_status = data.pivot_table(index = 'family_status', values = 'debt')
svod_status['quantity_status'] = data.groupby('family_status')['debt'].count()

def percent(x):
    return "{0:.2%}".format(x)

svod_status['debt'] = svod_status['debt'].apply(percent)
svod_status = svod_status.sort_values(by='debt', ascending=False)
svod_status.rename(columns={'debt':'Процент задолженостей', 'quantity_status':'Количество кредитов группе'})

Процент задолженостей  Количество кредитов группе
family_status                                                          
Не женат / не замужем                 9.76%                        2796
гражданский брак                      9.31%                        4134
женат / замужем                       7.56%                       12261
в разводе                             7.06%                        1189
вдовец / вдова                        6.62%                         951

**Вывод:** 

Больше всего должников среди групп: "Не женат/не замужме" и "гражданский брак" 9.76% и 9.31% соответсвенно. Самое низкое же количество задолженностей среди группы "вдовец/вдова", хотя количество случаем очень низкое в сравнение с общим количеством данных, поэтому именно в данном датасете считаю правильным считать, что самое низкое количество должников среди "женат/замужем" 7.56%, потому что выборка достаточна. 

<div class="alert alert-success">

**Артём:** 👍 Правильно.

</div>

#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

In [24]:
#0–30000 — 'E';
#30001–50000 — 'D';
#50001–200000 — 'C';
#200001–1000000 — 'B';
#1000001 и выше — 'A'.

svod_lvl_income = data.pivot_table(index = 'total_income_category', values = 'debt')
svod_lvl_income['quantity_income'] = data.groupby('total_income_category')['debt'].count()

def percent(x):
    return "{0:.2%}".format(x)

svod_lvl_income['debt'] = svod_lvl_income['debt'].apply(percent)
svod_lvl_income = svod_lvl_income.sort_values(by='debt', ascending=False)
svod_lvl_income.rename(columns={'debt':'Процент задолженостей', 'quantity_income':'Количество кредитов группе'})

Процент задолженостей  Количество кредитов группе
total_income_category                                                  
E                                     9.09%                          22
C                                     8.50%                       15921
A                                     8.00%                          25
B                                     7.06%                        5014
D                                     6.02%                         349

**Вывод:** 

Как мы видим из таблицы, категории уровня дохода "E" и "A" имеют очень низкое количество случаев, но можно сделать логичный вывод, что должников при уровне дохода "0-30000" наибольшее количество, потому что крайне низкая заработная плата. Если убрать две эти категории, то остаются "C" - средний уровень дохода, количество должников со средним уровнем дохода самое большое в предоставленных данных 8.5% из общего количества. И соответсвенно с ростом дохода уменьшается количество просроков "B" - 7.06%, группу "D" можно считать недостаточным количеством данных т.к доход является ниже среднего, а должников меньше всего 6.21%.

<div class="alert alert-info">

#### **Артём:** 💡 Дам тебе совет на будущее, как лучше проводить категоризацию количественных данных. **В этом проекте переделывать ничего не надо, так как заказчик уже предложил свои категории.** Если только как дополнительное исследование.
    
Сейчас самая маленькая категория в 728 раз меньше самой большой. Это может отразиться на результатах исследования. Когда мы делим на категории количественные данные, то лучше их делать примерно одинаковыми. Для этого можно воспользоваться методом `qcut()`. Пример:
    
```py
pd.qcut(df['total_income'], 4)
```
    
```
Name: total_income, Length: 21453, dtype: category
Categories (4, interval[float64]): [(20666.999, 107620.0]  (107620.0, 142594.0]  (142594.0, 195818.0]  (195818.0, 2265604.0]]
```
                                                                                                               
Что внутри: 
    
```py
pd.qcut(датафрейм['столбец'], количество_категорий)
``` 
    
Итого у нас 4 категории:
<ul>
  <li>меньше 107_620</li>
  <li>от 107_620 до 142_594</li>
  <li>от 142_594 до 195_818</li>
  <li>от 195_818</li>
</ul>

Спасибо большое, очень полезный метод, в курсе его нет, но так и правда информация для анализа становиться более явной!

<div class="alert alert-success">

**Артём:** 👍 Согласен.

</div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?

In [25]:
svod_lvl_income = data.pivot_table(index = 'purpose_category', values = 'debt')
svod_lvl_income['quantity_category'] = data.groupby('purpose_category')['debt'].count()

def percent(x):
    return "{0:.2%}".format(x)

svod_lvl_income['debt'] = svod_lvl_income['debt'].apply(percent)
svod_lvl_income = svod_lvl_income.sort_values(by='debt', ascending=False)
svod_lvl_income.rename(columns={'debt':'Процент задолженостей', 'quantity_category':'Количество кредитов группе'})

Процент задолженостей  Количество кредитов группе
purpose_category                                                          
операции с автомобилем                   9.35%                        4279
получение образования                    9.25%                        3988
проведение свадьбы                       7.91%                        2313
операции с недвижимостью                 7.26%                       10751

**Вывод:** 

Данные очень явные и показательные, потому что все группы имеют достаточное количество данных для анализа. Видно, что "операции с автомобилем" имеют больше всего должников 9.35%, скорее всего это связано с расходоми на машину и с процентами на автокредиты, они значительно выше чем на недвижимость. Несмотря на то, что количество кредитов на недвижимость самое большое, процент должников самый низкий 7.26%, этому способствует низкая ставка, плюс возможность взять кредит на квартиру до 30 лет, благодаря этому всему платеж будет небольшим.

<div class="alert alert-success">

**Артём:** 👍 Верно.

</div>

#### 3.5 Приведите возможные причины появления пропусков в исходных данных.

*Ответ:* 

Я думаю, что пропуски возникли из-за: 
1. Заемщики могли указать не все данные о себе намеренно, дабы скрыть информацию от банка.
2. Банку могла не понадобиться какая-то информация о заемщиках из-за уровня дохода, например, или же большого рабочего стажа.
3. Возможно еще кредит проходил по какой-либо специальной программе, что информацию туда была не нужна.

<div class="alert alert-success">

**Артём:** 👍 Хорошо.
  
</div> 

<div class="alert alert-success">

**Артём:**  А какие еще варианты?
    
</div>
<div class="alert alert-success">

**Артём:** 👍 Принято.

</div>

##### Дополнение по возможным причинам пропусков
4. Ошибки оператора, который заполнял анкету.
5. Банку могла не понадобиться информация о семейном положении или количестве детей из-за возроста, так же прибольшом первоначальном взносе уровень дохода можно не подтверждать. Таких вариаций может быть много, один из случаев описал в пункте 2.
6. А что если еще один и тот же человек брал несколько кредитов в одном банке, а полная информация заполнена только в строке об одном кредите.

#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.

*Ответ:* 

Потому что данные могут быть заполнены любым распределением, только при Гауссовском и равномерно распредленении количественных переменных стоит брать среднее значение, а при любом другом надежнее брать медианное значение, потому что большое количество данных будет распределено либо в большую, либо в меньшую сторону.
Наглядный пример:
(1, 3, 5, 7, 8, 10, 12, 14, 16) - среднее значение будет 8.4, а медианное 8, те можно брать любое из двух
(100, 100, 100, 100, 500) - среднее значение 180, а медианное 100, хорошо видно, что из-за одного завышенного значения, среднее сильно становиться больше, а, соответсвенно, заполнять пропуски, и проводить анализ данных с этим значение нельзя, тк данные будут неточными.

<div class="alert alert-success">

**Артём:** 👍 Отлично.

</div>

### Шаг 4: общий вывод.

Напишите ваш общий вывод.

По части проектной работы "Исследование данных", можно сделать вывод о человеке с большей вероятностью просрока кредита - это человек без детей, в браке, с доходом от 200001 до 1000000, который берет кредит на недвижимость. Эта категория будет самой низкой группой риска для банка.

Самой же высокой группой риска будет - человек имеющий 2х и более детей, не в браке, с доходом до 200000, который берет кредит на автомобиль.

<div class="alert alert-success">

**Артём:** 👍 Молодец.

</div>

## 🏁 Итоги ревью
<a id="results"></a>
[В начало](#start)

<div class="alert alert-danger">

**Артём:** ❌ Илья, спасибо за работу. Осталось исправить 1 комментарий и можно идти к заказчику. Молодец! 👍
    
Для удобства, комментарии которые нужно скорректировать я добавил в меню **Contents** слева.
    
Жду новую версию проекта. 🙂

</div>

Замечание исправлено!

<div class="alert alert-success">

**Артём:** 👍 Отлично.

</div>

<img src="https://lifeo.ru/wp-content/uploads/gif-salyut-10.gif"> 
<div class="alert alert-success">   

**Артём:** ✌️ Илья, спасибо за работу и внесенные корректировки! Критических замечаний нет.

Поздравляю, **проект принят!** 🥳 🎉 Ты молодец!

Желаю тебе успехов в дальнейшем обучении. 😉
    
---
    
#### 📖 **Полезные ссылки**

* [Работа в Яндексе](https://yandex.ru/jobs/pages/analytics_interview?ysclid=l84s63fs1g790141490)
* [Настройте свой Jupyter Notebook правильно](https://medium.com/nuances-of-programming/настройте-свой-jupyter-notebook-правильно-faef14c89434)    
* [Краткое руководство по Маркдауну](https://paulradzkov.com/2014/markdown_cheatsheet/?)
* [PEP 8 – Style Guide for Python Code](https://www.python.org/dev/peps/pep-0008/)    
* [Обработка пропусков в данных](https://loginom.ru/blog/missing)  
* [Разделение данных с помощью qcut и cut в Pandas](https://dfedorov.spb.ru/pandas/Разделение%20данных%20в%20Pandas%20с%20помощью%20qcut%20и%20cut.html?)  
* [Официальная документаци метода qcut](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.qcut.html)  
* [Подробное руководство по группировке и агрегированию с помощью pandas](https://dfedorov.spb.ru/pandas/Подробное%20руководство%20по%20группировке%20и%20агрегированию%20с%20помощью%20pandas.html?ysclid=lcmm6rtg2w602526506)
* [Шпаргалка по визуализации данных в Python с помощью Plotly](https://habr.com/ru/articles/502958/) 
* [Задачки SQL](https://habr.com/ru/company/dcmiran/blog/500360/)    
* [7 cайтов для оттачивания навыка написания SQL запросов на 2020 год](https://techrocks.ru/2019/11/22/7-sites-to-perfect-sql-query-writing-skills/)
* [Главред помогает очистить текст от словесного мусора, проверяет на соответствие информационному стилю.](https://glvrd.ru)

</div>